[processor/resourcedetection] Add k8s cluster name detection in EKS environment

crobert-1 opened the pull request:
**Description:** <Describe what has changed.>
<!--Ex. Fixing a bug - Describe the bug and how this fixes the issue.
Ex. Adding a feature - Explain what this achieves.-->
This enhancement detects the k8s cluster name in EKS. The solution uses EC2 instance tags to determine the cluster name, which means it will only work on EC2 (as noted in documentation updates).

Note: This is the second attempt at adding this functionality, superseding PR #27408. My first attempt was using `v2` of the [AWS GO SDK](https://github.com/aws/aws-sdk-go-v2), but the `service/ec2` module was causing lint to hit OOM issues and successful lint runs were taking over half an hour to complete. This uses `v1` of the module instead in an attempt to resolve linting issues.

**Link to tracking Issue:** <Issue number if applicable>
Resolves #26794

**Testing:** <Describe what testing was performed and which tests were added.>
Manual testing in an EKS cluster when running on EC2 instances. Added unit tests to cover basic functionality.

**Documentation:** <Describe the documentation added.>
Updated README and relevant metadata

crobert-1 committed:
Add k8s cluster name detection in EKS environment

Use ec2 instance tags to determine the EKS k8s cluster
name.

crobert-1 committed:
Remove longer EKS timeout as it's not necessary

dmitryax commented on processor/resourcedetectionprocessor/internal/aws/eks/detector.go:
```diff
@@ -74,7 +86,12 @@ func (d *detector) Detect(ctx context.Context) (resource pcommon.Resource, schem
+	// The error is unhandled because we want to return successfully detected resources
+	// regardless of an error. The caller will properly handle any error hit while getting
+	// the cluster name.
+	clusterName, err := d.utils.getClusterName(ctx)
+	d.rb.SetK8sClusterName(clusterName)
```
can you please wrap it in `if rb.config.K8sClusterName.Enabled {` so user don't run into errors issues if the atribute isn't needed

crobert-1 replied:
Good call, will do.

crobert-1 replied:
Actually, it looks like we don't have access to this since `config` is a private member of the `ResourceBuilder` struct in the `metadata` package. Is it worth re-architecting some of this to get access, or should we leave it as is?

dmitryax replied:
It should be easy to pass. Just add a new field to `detector` struct and set `cfg.ResourceAttributes` to it in `NewDetector` func

dmitryax commented on processor/resourcedetectionprocessor/internal/aws/eks/detector_test.go:
```diff
@@ -16,6 +17,11 @@ import (
 )
 
+const (
+	clusterNameKey = "k8s.cluster.name"
```
`conventions.AttributeK8SClusterName` can be used instead

dmitryax commented on processor/resourcedetectionprocessor/factory.go:
```diff
@@ -97,7 +101,7 @@ func createDefaultConfig() component.Config {
 func defaultHTTPClientSettings() confighttp.HTTPClientSettings {
 	httpClientSettings := confighttp.NewDefaultHTTPClientSettings()
-	httpClientSettings.Timeout = 5 * time.Second
+	httpClientSettings.Timeout = defaultTimeout
```
please avoid unrelated changes

dmitryax commented on processor/resourcedetectionprocessor/README.md:
```diff
@@ -334,21 +338,21 @@ processors:
-[Function as a Service semantic conventions](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/resource/semantic_conventions/faas.md)
-and [AWS Lambda semantic conventions](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/instrumentation/aws-lambda.md#resource-detector)
+[Function as a Service semantic conventions](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/resource/faas.md)
+and [AWS Lambda semantic conventions](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/faas/aws-lambda.md)
```
These fixes would be nice as a separate PR

crobert-1 replied:
Moved to #28669.

dmitryax commented on processor/resourcedetectionprocessor/internal/aws/eks/detector.go:
```diff
@@ -114,3 +131,49 @@ func (e eksDetectorUtils) getConfigMap(ctx context.Context, namespace string, na
 }
+
+func (e eksDetectorUtils) getClusterName(ctx context.Context) (string, error) {
+	sess := session.Must(session.NewSession())
```
Can we avoid panic and return an error instead?

crobert-1 committed:
Changes requested by Dmitrii

- Don't panic on error
- Remove unrelated changes
- Reuse conventions variable for k8s cluster name

crobert-1 commented:
`check-links` failure will be fixed by #28669

crobert-1 committed:
Merge branch 'main' into eks_cluster_name_sdk_v1

crobert-1 committed:
Only attempt to get cluster name if it's enabled

dmitryax commented:
@jinja2 can you please take a look as well?

dmitryax commented on processor/resourcedetectionprocessor/internal/aws/eks/detector.go:
```diff
@@ -114,3 +137,53 @@ func (e eksDetectorUtils) getConfigMap(ctx context.Context, namespace string, na
+	for _, reservation := range reservations {
+		for _, instance := range reservation.Instances {
+			for _, tag := range instance.Tags {
+				if *tag.Key == clusterNameAwsEksTag || *tag.Key == clusterNameEksTag {
```
Is the key guaranteed to be not `nil`?

crobert-1 replied:
I just checked, there's no enforcement in place to guarantee it's non-nil. I'll add a check here. 👍

crobert-1 committed:
Add nil check for tag's key

jinja2 commented on processor/resourcedetectionprocessor/README.md:
```diff
@@ -318,14 +318,18 @@ processors:
   resourcedetection/eks:
     detectors: [env, eks]
-    timeout: 2s
+    timeout: 15s
```
Is this the avg duration you noticed in your tests? I assume users will be mindful of this additional delay and update readiness check timeouts when the healthcheck extension + k8s probe are enabled.

crobert-1 replied:
I was generally testing around `15s`, I think around `11s` was the lowest timeout that usually worked before timing out.

Do you think a comment should be added here regarding using this with the healthcheck extension?

jinja2 replied:
Just checked the liveness probe values set by default in the helm-chart, we should be good if 15s is the worst observed (probe is run 3 times with 10sec interval before restarting). I was hoping the [healthcheck extension](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/extension/healthcheckextension) had some documentation about things to consider for setting intervals/threshold for health probes, but no dice.

bryan-aguilar commented on processor/resourcedetectionprocessor/internal/aws/eks/detector.go:
```diff
@@ -74,7 +88,16 @@ func (d *detector) Detect(ctx context.Context) (resource pcommon.Resource, schem
 	d.rb.SetCloudPlatform(conventions.AttributeCloudPlatformAWSEKS)
 
-	return d.rb.Emit(), conventions.SchemaURL, nil
+	// The error is unhandled because we want to return successfully detected resources
+	// regardless of an error. The caller will properly handle any error hit while getting
+	// the cluster name.
+	if d.ra.K8sClusterName.Enabled {
+		var clusterName string
+		clusterName, err = d.utils.getClusterName(ctx)
+		d.rb.SetK8sClusterName(clusterName)
+	}
+
+	return d.rb.Emit(), conventions.SchemaURL, err
```
Who is the caller in this case and how is it handled? Is the caller built in such a way where it treats this as a partial error and doesn't immediately exit on error?

crobert-1 replied:
Good point, caller is here:
https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/b716a4d89acd497cb1ed044b941c49a7897b72bb/processor/resourcedetectionprocessor/internal/resourcedetection.go#L129-L134

The error will result in the correctly detected resources not being returned. I'll add a warning log message instead of returning an error here, I believe that's the best way to balance this.

crobert-1 replied:
Tested with update, here's the logging output:
```
2023-11-09T23:03:04.703Z	info	internal/resourcedetection.go:125	began detecting resource information	{"kind": "processor", "name": "resourcedetection/eks", "pipeline": "logs"}
2023-11-09T23:03:07.846Z	warn	eks/detector.go:97	Unable to get EKS cluster name	{"kind": "processor", "name": "resourcedetection/eks", "pipeline": "logs", "error": "EC2MetadataRequestError: failed to get EC2 instance identity document\ncaused by: RequestCanceled: request context canceled\ncaused by: context deadline exceeded"}
2023-11-09T23:03:07.846Z	info	internal/resourcedetection.go:139	detected resource information	{"kind": "processor", "name": "resourcedetection/eks", "pipeline": "logs", "resource": {"cloud.platform":"aws_eks","cloud.provider":"aws"}}
```
We are able to receive successfully detected resources, with the log message for the `k8s.cluster.name` failure.

crobert-1 committed:
Handle error when getting k8s cluster name

bryan-aguilar commented on processor/resourcedetectionprocessor/internal/aws/eks/detector.go:
```diff
@@ -74,7 +88,20 @@ func (d *detector) Detect(ctx context.Context) (resource pcommon.Resource, schem
 	d.rb.SetCloudPlatform(conventions.AttributeCloudPlatformAWSEKS)
 
-	return d.rb.Emit(), conventions.SchemaURL, nil
+	if d.ra.K8sClusterName.Enabled {
```
Do we need to perform this check ourselves? It looks like the generated code does that for you inside `SetK8sClusterName`. I could see the benefit of doing the check, that way we only perform the query if it's enabled. It could confuse users if the see an error for cluster name but they don't have it enabled.

crobert-1 replied:
Yes, you're hitting on the reasons why we added this check explicitly. The logic to get the k8s cluster name can take 10+ seconds, and will often hit errors when running with the default timeout (or always hit errors if user doesn't have proper permission). Adding this means the user won't have the long delay and unrelated errors if the cluster name is disabled.

bryan-aguilar commented on processor/resourcedetectionprocessor/internal/aws/eks/detector.go:
```diff
@@ -74,7 +88,20 @@ func (d *detector) Detect(ctx context.Context) (resource pcommon.Resource, schem
+	if d.ra.K8sClusterName.Enabled {
+		var clusterName string
+		clusterName, err = d.utils.getClusterName(ctx)
+		if err != nil {
+			// We clear out the error here so that the correctly detected resources
+			// can still be used by the processor.
+			d.logger.Warn("Unable to get EKS cluster name", zap.Error(err))
+			err = nil
+		} else {
+			d.rb.SetK8sClusterName(clusterName)
+		}
```
Is this going to log each time a datapoint is processed or only once on startup? If it's each time then we should consider making this debug and not warn level. 

Also, manually setting the error here feels like an anti pattern. If the error is only used for a logging statement, and never returned to the caller, would it make more sense to not return an error at all from `getClusterName`? A logging statement could just be added inside that function.

crobert-1 replied:
This is only logged once at collector startup.

Example logs: (collector is running with filelog receiver and debug exporter)
```
$ kubectl -n crobert-test logs otel-collector-7985f54766-dtdbf
2023-11-10T16:55:22.020Z	info	service@v0.88.1-0.20231026220224-6405e152a2d9/telemetry.go:84	Setting up own telemetry...
2023-11-10T16:55:22.020Z	info	service@v0.88.1-0.20231026220224-6405e152a2d9/telemetry.go:201	Serving Prometheus metrics	{"address": ":8888", "level": "Basic"}
2023-11-10T16:55:22.020Z	info	exporter@v0.88.1-0.20231026220224-6405e152a2d9/exporter.go:275	Development component. May change in the future.	{"kind": "exporter", "data_type": "logs", "name": "debug"}
2023-11-10T16:55:22.110Z	info	service@v0.88.1-0.20231026220224-6405e152a2d9/service.go:143	Starting otelcontribcol...	{"Version": "0.88.0-dev", "NumCPU": 16}
2023-11-10T16:55:22.110Z	info	extensions/extensions.go:33	Starting extensions...
2023-11-10T16:55:22.110Z	info	internal/resourcedetection.go:125	began detecting resource information	{"kind": "processor", "name": "resourcedetection/eks", "pipeline": "logs"}
2023-11-10T16:55:25.310Z	warn	eks/detector.go:97	Unable to get EKS cluster name	{"kind": "processor", "name": "resourcedetection/eks", "pipeline": "logs", "error": "EC2MetadataRequestError: failed to get EC2 instance identity document\ncaused by: RequestCanceled: request context canceled\ncaused by: context deadline exceeded"}
2023-11-10T16:55:25.310Z	info	internal/resourcedetection.go:139	detected resource information	{"kind": "processor", "name": "resourcedetection/eks", "pipeline": "logs", "resource": {"cloud.platform":"aws_eks","cloud.provider":"aws"}}
2023-11-10T16:55:25.310Z	info	adapter/receiver.go:45	Starting stanza receiver	{"kind": "receiver", "name": "filelog", "data_type": "logs"}
2023-11-10T16:55:25.310Z	info	service@v0.88.1-0.20231026220224-6405e152a2d9/service.go:169	Everything is ready. Begin running and processing data.
2023-11-10T16:55:25.511Z	info	fileconsumer/file.go:251	Started watching file	{"kind": "receiver", "name": "filelog", "data_type": "logs", "component": "fileconsumer", "path": "/var/log/pods/analytics-tests-rv_quantizer-alpha-0_551d1187-3e40-4179-a5fb-3b752da44ad6/quantizer/0.log"}
2023-11-10T16:55:28.811Z	info	LogsExporter	{"kind": "exporter", "data_type": "logs", "name": "debug", "resource logs": 1, "log records": 5}
2023-11-10T16:55:32.611Z	info	LogsExporter	{"kind": "exporter", "data_type": "logs", "name": "debug", "resource logs": 1, "log records": 2}
2023-11-10T16:55:41.810Z	info	LogsExporter	{"kind": "exporter", "data_type": "logs", "name": "debug", "resource logs": 2, "log records": 3}
2023-11-10T16:55:44.610Z	info	LogsExporter	{"kind": "exporter", "data_type": "logs", "name": "debug", "resource logs": 1, "log records": 2}
2023-11-10T16:55:50.211Z	info	LogsExporter	{"kind": "exporter", "data_type": "logs", "name": "debug", "resource logs": 1, "log records": 2}
2023-11-10T16:56:00.211Z	info	LogsExporter	{"kind": "exporter", "data_type": "logs", "name": "debug", "resource logs": 2, "log records": 6}
2023-11-10T16:56:01.210Z	info	LogsExporter	{"kind": "exporter", "data_type": "logs", "name": "debug", "resource logs": 2, "log records": 4}
2023-11-10T16:56:02.811Z	info	LogsExporter	{"kind": "exporter", "data_type": "logs", "name": "debug", "resource logs": 1, "log records": 10}
2023-11-10T16:56:04.011Z	info	LogsExporter	{"kind": "exporter", "data_type": "logs", "name": "debug", "resource logs": 2, "log records": 6}
2023-11-10T16:56:05.011Z	info	LogsExporter	{"kind": "exporter", "data_type": "logs", "name": "debug", "resource logs": 2, "log records": 4}
2023-11-10T16:56:06.713Z	info	LogsExporter	{"kind": "exporter", "data_type": "logs", "name": "debug", "resource logs": 1, "log records": 100}
2023-11-10T16:56:06.811Z	info	LogsExporter	{"kind": "exporter", "data_type": "logs", "name": "debug", "resource logs": 1, "log records": 14}
2023-11-10T16:56:08.611Z	info	LogsExporter	{"kind": "exporter", "data_type": "logs", "name": "debug", "resource logs": 1, "log records": 1}
2023-11-10T16:56:08.810Z	info	LogsExporter	{"kind": "exporter", "data_type": "logs", "name": "debug", "resource logs": 2, "log records": 3}
2023-11-10T16:56:11.411Z	info	LogsExporter	{"kind": "exporter", "data_type": "logs", "name": "debug", "resource logs": 3, "log records": 7}
2023-11-10T16:56:11.811Z	info	LogsExporter	{"kind": "exporter", "data_type": "logs", "name": "debug", "resource logs": 1, "log records": 6}
2023-11-10T16:56:15.211Z	info	LogsExporter	{"kind": "exporter", "data_type": "logs", "name": "debug", "resource logs": 2, "log records": 7}
2023-11-10T16:56:16.211Z	info	LogsExporter	{"kind": "exporter", "data_type": "logs", "name": "debug", "resource logs": 2, "log records": 4}
```
I agree the manual setting of `err = nil` should be improved. I think it would be fair to log each error in `getClusterName` at the warn level, and simply return an empty string. No harm is done by setting the k8s.cluster.name to be empty, so I that sounds like a fair solution to me.

crobert-1 committed:
Refactor around error handling and logging

bryan-aguilar commented on processor/resourcedetectionprocessor/internal/aws/eks/detector.go:
```diff
@@ -74,7 +88,12 @@ func (d *detector) Detect(ctx context.Context) (resource pcommon.Resource, schem
+		d.rb.SetK8sClusterName(clusterName)
+	}
+
+	return d.rb.Emit(), conventions.SchemaURL, err
```
```suggestion
	return d.rb.Emit(), conventions.SchemaURL, nil
```

The only reason I am suggesting this is that the `isEKS()` doesn't actually handle the error directly. I'm not sure if it's possible but there could be a scenario where `isEKS == true` and `err != nil`. In this case you would be returning an error here and the resources would be dropped by caller. 

I think we could return nil here and/or check for non nil error above.

crobert-1 replied:
Fair point. There currently isn't any scenario where `err` would be non-nil, and `isEKS` would be true, but I could see the confusion.

crobert-1 committed:
Update processor/resourcedetectionprocessor/internal/aws/eks/detector.go

Co-authored-by: bryan-aguilar <46550959+bryan-aguilar@users.noreply.github.com>

crobert-1 commented:
Lint failure unrelated (fixed by #29101), signalfx exporter test failure is also unrelated (added frequency to #29099)

crobert-1 committed:
Merge branch 'main' into eks_cluster_name_sdk_v1

crobert-1 committed:
Merge branch 'main' into eks_cluster_name_sdk_v1
